target/linux/ath79/files/drivers/mtd/nand/raw/rb4xx-nand.c

```diff
@@ -0,0 +1,316 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * NAND driver for the MikroTik RouterBoard 4xx series
+ *
+ * Copyright (C) 2008-2011 Gabor Juhos <juhosg@openwrt.org>
+ * Copyright (C) 2008 Imre Kaloz <kaloz@openwrt.org>
+ * Copyright (C) 2015 Bert Vermeulen <bert@biot.com>
+ * Copyright (C) 2020 Christopher Hill <ch6574@gmail.com>
+ *
+ * This file was based on the driver for Linux 2.6.22 published by
+ * MikroTik for their RouterBoard 4xx series devices.
+ */
+#include <linux/platform_device.h>
+#include <linux/mtd/rawnand.h>
+#include <linux/gpio/consumer.h>
+#include <linux/module.h>
+#include <linux/of_device.h>
+#include <linux/version.h>
+
+#include <mfd/rb4xx-cpld.h>
+
+struct rb4xx_nand {
+	struct rb4xx_cpld *cpld;
+	struct device *dev;
+
+	struct nand_chip chip;
+	struct gpio_desc *ale;
+	struct gpio_desc *cle;
+	struct gpio_desc *nce;
+	struct gpio_desc *rdy;
+};
+
+static int rb4xx_ooblayout_ecc(struct mtd_info *mtd, int section,
+				struct mtd_oob_region *oobregion)
+{
+	switch (section) {
+	case 0:
+		oobregion->offset = 8;
+		oobregion->length = 3;
+		return 0;
+	case 1:
+		oobregion->offset = 13;
+		oobregion->length = 3;
+		return 0;
+	default:
+		return -ERANGE;
+	}
+}
+
+static int rb4xx_ooblayout_free(struct mtd_info *mtd, int section,
+				struct mtd_oob_region *oobregion)
+{
+	switch (section) {
+	case 0:
+		oobregion->offset = 0;
+		oobregion->length = 4;
+		return 0;
+	case 1:
+		oobregion->offset = 4;
+		oobregion->length = 1;
+		return 0;
+	case 2:
+		oobregion->offset = 6;
+		oobregion->length = 2;
+		return 0;
+	case 3:
+		oobregion->offset = 11;
+		oobregion->length = 2;
+		return 0;
+	default:
+		return -ERANGE;
+	}
+}
+
+static const struct mtd_ooblayout_ops rb4xx_nand_ecclayout_ops = {
+	.ecc = rb4xx_ooblayout_ecc,
+	.free = rb4xx_ooblayout_free,
+};
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,20,0)
+static uint8_t rb4xx_nand_read_byte(struct mtd_info *mtd)
+{
+	struct rb4xx_nand *nand = mtd->priv;
+#else
+static u8 rb4xx_nand_read_byte(struct nand_chip *chip)
+{
+	struct rb4xx_nand *nand = chip->priv;
+#endif
```

Comment:
Yuck. Now I see what you meant! I've seen worse though :)
Note that u8 == uint8_t

Reply:
Indeed, I was half thinking maybe I can just wait until this all flips to 5.x kernel and let it go away!

Reply:
BTW, the `#if` sections are all for the api changes moving away from `mtd_info` and to `nand_chip` plus the newly embedded `legacy` struct

Reply:
I don't know what's the plan for switching ath79 over to 5.4 (maybe @adrianschmutzler or @xback can comment on that) but if it's going to happen soon then yes, it makes sense to leave these out.

Reply:
Even if we switch soon, I expect 4.19 to be around after that for quite some time, and wouldn't want to remove the compatibility without need. Personally, I don't think the conditions are that terrible, and would just live with them. An alternative option would be to use kernel patches instead of files for these changes, where we wouldn't need version switches at all, but have additional duplicate code and less comfort for reading the parts without switches.

I personally am quite 50:50 between both versions, so choose as you wish.

Reply:
Just to confirm, 4.19 will stay active for new ath79 targets even after switching to 5.4 ?

Reply:
Well, what's "active"?

It has been considered convenient in the past to have an older (known-to-be-good) kernel version available for testing, to narrow down regressions, etc.

Nobody will build from it by default, but one can if it's helpful. The same was done for 4.19 and 4.14 in parallel on many targets for quite some time.

Reply:
BTW, since only the nand driver is affected by these switches, you could also just implement the nand driver as kernel patches, and keep the rest as files, taking the best of both worlds. But still, it's only up to you what you prefer.

Reply:
If I may chime in on this, I really wound't go that way: these files are in fact one cohesive driver, they should be seen as a one driver; splitting the nand driver in separate patches for each kernel version duplicates code and review work (and reviewing patches of patches is unpleasant); and the #defines are sufficiently sparse to remain manageable. If/when this driver ends up in mainline we can handle a small backport patch locally, but I really wouldn't do this now. My quick 2c (I'll review the last push later).

Reply:
I'm going to leave the `#if` blocks in as on reflection they might look ugly but they are easy to follow (however any other nand drivers people may create for this dual 4.19 - 5.4 world will run into the same issue of API breakage).

N.B. I also plan to rename the .c files, from `rb4xx-gpio.c` to `gpio-rb4xx.c` and so on as that seems to be the prevalent pattern used upstream.